src/utils/url.js

```diff
@@ -67,11 +90,40 @@ const mimes = {
 export const getMimeTypeFromFileExtension = (extension: string): string =>
   mimes[extension.toLowerCase()] || 'application/octet-stream';
 
-export const autocompleteUrl = (value: string = '', protocol: string, append: string): string =>
-  value.length > 0
-    ? `${hasProtocol(value) ? '' : protocol}${value || 'your-org'}${
-        value.indexOf('.') === -1 ? append : ''
-      }`
-    : '';
-
 export const isValidUrl = (url: string): boolean => urlRegex({ exact: true }).test(url);
+
+export type AutocompletionData = {|
+  protocol: string,
+  hostname: string,
+|};
+
+export type AutocompletionPieces = [string | null, string, string | null];
+
+/**
+ * Given user input purporting to be a Zulip server's host name, provide
+ * a prefix, derived value, and suffix which may suffice to turn it into
+ * a real URL.
+ *
+ * Presently, the derived value will always be equal to the input value;
+ * this should not be relied on, as it may change in future.
+ */
+export const autocompleteHostnamePieces = (
+  value: string,
+  data: AutocompletionData,
+): AutocompletionPieces => {
+  const [protocol, nonProtocolValue] = parseProtocol(value);
+
+  const prefix: string | null = protocol === null ? data.protocol : null;
+
+  const suffix: string | null =
+    nonProtocolValue.includes('.') || nonProtocolValue.includes(':') ? null : `.${data.hostname}`;
```

Comment:
The behavior change on `:` here would be great to pull out as a (very small) commit of its own, coming after the refactor.

If you do, and also the small other behavior change I mentioned above, does the main commit become a completely pure refactor? If so, that's extra helpful.